Added CreateAllelicPanelOfNormals.

samuelklee opened the pull request:
Also:
-Changed final static -> static final.
-Minor changes and cleanup.
-CoveragePanelOfNormals interfaces and compatible classes.
-Major refactoring of PCA coverage PoN.
-Test files for allelic PoN.

samuelklee commented:
@asmirnov239, can you review the 4 non-coverage-PoN-related commits?  Feel free to comment on anything that is unclear, and check for violations of the Java style guide (https://google.github.io/styleguide/javaguide.html).  Thanks!

samuelklee commented:
Oh, and one thing to note: I left a few TODOs in places where I might potentially have merge conflicts or will make changes after #596 gets merged.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/exome/allelefraction/AlleleFractionInitializer.java:
```diff
@@ -166,4 +166,55 @@ private double estimateMinorFraction(final int segment, final AlleleFractionData
                 .mapToDouble(segment -> estimateMinorFraction(segment, data))
                 .boxed().collect(Collectors.toList()));
     }
+
```
Just putting a note here that this block of code used to live in `AllelicPanelOfNormals` and I just moved it into this class (largely unchanged).  It will probably need updating and might get shuffled around again after #596 goes in.

davidbenjamin commented:
I believe we agreed at standup that @asmirnov239 will review the first four commits and I will review the last two.  Do I have that right?

samuelklee commented:
@davidbenjamin if you could review "CoveragePanelOfNormals interfaces and compatible classes" and "Major refactoring of PCA coverage PoN" that'd be great!

davidbenjamin commented:
@samuelklee got it!

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNNormalizationResult.java:
```diff
@@ -0,0 +1,10 @@
+ * Interface for a coverage profile that has been normalized by a {@link CoveragePanelOfNormals}.
+ */
+public interface CoveragePoNNormalizationResult {
+    ReadCountCollection getProfile();
```
I think you can give this the `@FunctionalInterface` annotation, which would keep the compiler from complaining if anyone ever wanted to instantiate the interface as a lambda.

samuelklee replied:
will add it for now, but I wonder if we will want to add any methods to this interface in the future?

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNQCUtils.java:
```diff
@@ -24,47 +23,41 @@
+        final Map<String, Double> contigToMedian = getContigToMedianCRMap(qcNormalization.getProfile());
+
+        // Segment the result
+        // For each tangent normalization result from the QC PoN
```
Can this comment be on a single line?

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNQCUtils.java:
```diff
@@ -24,47 +23,41 @@
+        // For each tangent normalization result from the QC PoN
+        if (ctx == null) {
+            final List<ReadCountCollection> singleSampleTangentNormalizedReadCounts = createIndividualReadCountCollections(qcNormalization.getProfile());
+            result = identifySamplesWithSuspiciousContigs(singleSampleTangentNormalizedReadCounts, contigToMedian);
```
Instead of declaring `result` without initializing above, why not just `return identifySamplesWith. . .` here and below?

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNQCUtils.java:
```diff
@@ -115,14 +108,13 @@ private static Boolean hasSuspiciousContigs(final ReadCountCollection singleSamp
     static List<String> identifySamplesWithSuspiciousContigs(final List<ReadCountCollection> singleSampleTangentNormalizedReadCounts, final Map<String, Double> contigToMedian) {
         logger.info("Checking for samples with suspicious contigs...");
-        return singleSampleTangentNormalizedReadCounts.stream().filter(r -> PoNIssueDetector.hasSuspiciousContigs(r, contigToMedian)).map(r -> r.columnNames().get(0)).collect(Collectors.toList());
+        return singleSampleTangentNormalizedReadCounts.stream().filter(r -> hasSuspiciousContigs(r, contigToMedian)).map(r -> r.columnNames().get(0)).collect(Collectors.toList());
```
Slight preference for the lambda variable to be `rcc` for `ReadCountCollection`.

samuelklee replied:
don here and below

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNQCUtils.java:
```diff
@@ -24,47 +23,41 @@
+            final List<ReadCountCollection> singleSampleTangentNormalizedReadCounts = createIndividualReadCountCollections(qcNormalization.getProfile());
+            result = identifySamplesWithSuspiciousContigs(singleSampleTangentNormalizedReadCounts, contigToMedian);
+        } else {
+            final JavaRDD<ReadCountCollection> parallelSingleSampleTangentNormalizedReadCounts = createParallelIndividualReadCountCollections(qcNormalization.getProfile(), ctx);
```
I couldn't find a common superinterface for `List` and `JavaRDD` that would unify the parallel and non-parallel code, but if you know of something that would be great.

samuelklee replied:
Not sure either!

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/CoveragePoNQCUtils.java:
```diff
@@ -154,6 +146,6 @@ private static Boolean hasSuspiciousContigs(final ReadCountCollection singleSamp
 
     private static List<String> retrieveAllContigsPresent(final ReadCountCollection readCountCollection) {
-        return readCountCollection.targets().stream().map(Target :: getContig).distinct().collect(Collectors.toList());
+        return readCountCollection.targets().stream().map(Target:: getContig).distinct().collect(Collectors.toList());
```
space after ::

samuelklee replied:
oops, fixed one space but not the other!

davidbenjamin commented:
Done with review of "CoveragePanelOfNormals interfaces and compatible classes."

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -38,8 +39,8 @@
 @CommandLineProgramProperties(
-        summary = "Create a Panel of Normals (PoN) given the proportional read counts for the samples that are part of the panel.  Supports Apache Spark for some operations.",
-        oneLineSummary = "Create a Panel of Normals",
+        summary = "Create a coverage panel of normals (PoN) given the proportional read counts for the samples that are part of the panel.  Supports Apache Spark for some operations.",
```
"the samples that are part of the panel" -> "samples in the panel"

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -48,30 +49,21 @@
-
-    public static final double DEFAULT_MAXIMUM_PERCENT_ZEROS_IN_TARGET = 5.0;
-
     public static final String INFER_NUMBER_OF_EIGEN_SAMPLES = "auto";
```
Not your code, but seeing "eigen samples" spelled as two words pains me.

davidbenjamin replied:
Probably not worth fixing since the new coverage model won't have eigensamples or eigen samples.

samuelklee replied:
fixed camel case, spacing, and underscores everywhere, except in HDF5 path `PCA/EIGEN_VECTORS` in `PCA` class

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -229,47 +229,46 @@ protected void runPipeline(JavaSparkContext ctx) {
         }
 
         // Check parameters and load values to meet the backend PoN creation interface
+        validateArguments();
```
much better

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -229,47 +229,46 @@ protected void runPipeline(JavaSparkContext ctx) {
+                final List<String> failingSampleNames = CoveragePoNQCUtils.retrieveSamplesWithArmLevelEvents(qcPoN, ctx);
+                ParamUtils.writeStringListToFile(failingSampleNames, blacklistOutFile);
+                // If no suspicious samples were found, just redo the PoN reduction to save time.
+                if (failingSampleNames.size() != 0) {
```
Use `isEmpty()`

samuelklee replied:
fixed everywhere

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -229,47 +229,46 @@ protected void runPipeline(JavaSparkContext ctx) {
+                    HDF5PCACoveragePoNCreationUtils.create(ctx, outFile, HDF5File.OpenMode.CREATE, inputFile, targets, failingSampleNames,
+                            targetFactorThreshold, maximumPercentZerosInColumn, maximumPercentZerosInTarget,
+                            columnExtremeThresholdPercentile, outlierTruncatePercentileThresh, numberOfEigenSamples, dryRun
+                    );
```
`);` belongs on the previous line

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreatePanelOfNormals.java:
```diff
@@ -279,92 +278,54 @@ protected void runPipeline(JavaSparkContext ctx) {
-            final HDF5PoN pon = new HDF5PoN(file);
-            writeTargetWeightsFile(pon, outputFile);
+            final HDF5PCACoveragePoN pon = new HDF5PCACoveragePoN(file);
+            final double[] targetWeights = DoubleStream.of(pon.getTargetVariances()).boxed().mapToDouble(v -> 1/v).toArray();
```
Isn't `.boxed().mapToDouble()` equivalent to `.map()`?

davidbenjamin replied:
FYI it's not necessary here but for performance-sensitive code you could avoid constructing a stream with `MathUtils.applyToArray(pon.getTargetVariance(), v -> 1/v)`.

samuelklee replied:
oops, good catch!  not sure if the public dependency has been updated yet in my branch, so I don't see `applyToArray`, will check again after a rebase

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+ * HDF5 File backed coverage panel of normals data structure.
+ *
+ * Currently, we assume the coverage panel of normals is always fully specified but that the allelic panel of normals is
+ * essentially optional.
```
What does "essentially" mean?

samuelklee replied:
oops, forgot to remove this comment, thanks for catching it!  (originally i tried to add the aPoN as an optional field of the coverage PoN, but scrapped that for various reasons)

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+ * Several attributes are stored transposed (in other words, the rows and columns are interchanged).
+ * This dodges a very slow write time in HDF5, since we usually have many more rows (targets) than columns (samples).
+ *
+ * HDF5 can write matrices with few rows and many columns much faster than matrices with many rows and few columns.
```
can write -> writes

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+     * The version number is a float point number (double) where the integer part is the
+     * major and the decimal part is the minor.
+     *
+     * Note: the logic choice is to use a free form string but, the Python version was using
```
logic > logical

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+    /**
+     * The version number.
+     *
+     * The version number is a float point number (double) where the integer part is the
```
float -> floating

samuelklee replied:
just did "float point number (double) -> double"

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+    @Override
+    public RealMatrix getLogNormalizedPInverseCounts() {
+        return file.readMatrixAndCheckDimensions(LOG_NORMALIZED_PINV_PATH, getPanelSampleNames().size(),
+                getPanelTargetNames().size());
```
Why no transpose here?

davidbenjamin replied:
I'm guessing it's because pseudoinverses switch the row and column dimensionality.  If that's the reason, it deserves a comment.

samuelklee replied:
right, this is mentioned in the javadoc for the class but i can add comments here.

also, just realized that some of the dimensions in the PCACoveragePoN javadoc are also wrong, will fix these as well.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    @Override
+    public RealMatrix getReducedPanelPInverseCounts() {
```
Same need for a comment about the lack of transpose.

samuelklee replied:
see above

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNCreationUtils.java:
```diff
@@ -31,298 +30,198 @@
+                                                                               final double targetFactorPercentileThreshold, final Logger logger) {
+        final double[] targetFactors = calculateTargetFactors(readCounts);
+        final Percentile percentile = new Percentile(targetFactorPercentileThreshold);
+        final double threshold = percentile.evaluate(targetFactors);
```
I would find it clearer if `threshold` were on the LHS and `targetFactorPercentileThreshold` on the RHS via `final double threshold = new Percentile(targetFactorPercentileThreshold).evaluate(targetFactors);`

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNCreationUtils.java:
```diff
@@ -350,150 +249,151 @@ public double visit(final int row, final int column, final double value) {
+    static ReadCountCollection removeColumnsWithTooManyZeros(final ReadCountCollection readCounts, final int maximumColumnZeros, final Logger logger) {
+        final RealMatrix counts = readCounts.counts();
+
+        final Set<String> columnsToKeep = IntStream.range(0, counts.getColumnDimension()).boxed()
```
Instead of `.boxed() . . . map()` how about skip the `boxed()` and use `mapToObj()`?

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNCreationUtils.java:
```diff
@@ -531,178 +430,173 @@ static ReadCountCollection removeTargetsWithTooManyZeros(final ReadCountCollecti
-                    String.format("No count column dropped due to zero counts; there is no column with a large number of targets with zero counts (<= %d of %d) ", maximumColumnZeros, readCounts.targets().size()));
+        // Determine kept and dropped column sets.
+        final Set<String> columnsToKeep = new HashSet<>(readCounts.columnNames().size());
+        final Map<String, Double> columnsToDrop = new LinkedHashMap<>((int) (extremeColumnMedianCountPercentileThreshold * 4.0 / 100.0));
```
Where does the `percentileThreshold * 4/100` capacity come from?  I would have thought it was `2 * (percentileThreshold/100) * readCounts.columnNames().size()`.

samuelklee replied:
hmm, not sure.  although it probably doesn't make a huge difference, will change it as you suggest.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNCreationUtils.java:
```diff
@@ -531,178 +430,173 @@ static ReadCountCollection removeTargetsWithTooManyZeros(final ReadCountCollecti
-    static double[] calculateRowVariances(final RealMatrix m) {
-        Utils.nonNull(m, "Cannot calculate variances for a null matrix.");
-        return MatrixSummaryUtils.getRowVariances(m);
+        for (int i = 0; i < targetCount; i++) {
```
Can this be implemented with a non-modifying matrix `Visitor`?

samuelklee replied:
perhaps, but i think this is beyond the scope of this PR.  @mbabadi might look into adding it to #634, but perhaps the same could be said for that PR as well.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+    private static final String NUM_TARGET_COLUMNS_GROUP_NAME = "/num_target_cols";
+    private static final String NUM_TARGET_COLUMNS_PATH = NUM_TARGET_COLUMNS_GROUP_NAME + "/values";
+
+    private static final Map<String, Integer> targetColumnToPoNIndex = ImmutableMap.of(
```
Can this be implemented as an `EnumMap<TargetTableColumn, Integer>`?

samuelklee replied:
done here and for snpColumnToPoNIndex

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -0,0 +1,483 @@
+        final String[][] targetValues = new String[targets.size()][NUM_TARGET_COLUMNS];
+        for (int i = 0; i < targets.size(); i++) {
+            final Target target = targets.get(i);
+            targetValues[i][targetColumnToPoNIndex.get(TargetTableColumn.CONTIG.toString())] = target.getContig();
```
See above about implementing `targetColumnToPoNIndex` as an `EnumMap`

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationResult.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    /**
+     * Writes the tangent-factor normalized outputs.
```
Old terminology?

samuelklee replied:
good catch, done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationResult.java:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Writes the tangent-factor normalized outputs.
+     *
+     * Please note that this file is written in the target format originally used in recapseg.  In other words, it is
```
Better just to say "outputs in the format described in {@link ___}`

samuelklee replied:
OK, used `{@link ReadCountCollectionUtils}`

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationResult.java:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Writes the pre-tangent-normalization read counts if a file was provided for it.
+     *
+     * This file is written in the target format originally used in recapseg.  In other words, it is
```
likewise

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationResult.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private void writeOutput(final File file, final ReadCountCollection counts, final String commandLine, final String title) {
+        if (file == null) {
```
when would someone try to write to a `null` file?  Should this throw an exception instead?

samuelklee replied:
writing to these files is optional, in which case they are set to null by the calling class.  added to the javadoc to clarify

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+     * </p>
+     */
+    public static double EPSILON = 1E-9;
+    private static double INV_LN2 = 1.0 / Math.log(2.0);
```
If this isn't in `MathUtils` or `GATKProtectedMathUtils` it belongs there.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    static void factorNormalize(final RealMatrix input, final double[] targetFactors) {
+        Utils.nonNull(input, "Input matrix cannot be null.");
+        Utils.nonNull(targetFactors, "Target factors cannot be null.");
+        if (targetFactors.length != input.getRowDimension()) {
```
I'm comfortable with throwing an `IllegalArgumentException` for this, which lets you use `Utils.validateArg`.

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    private PCATangentNormalizationUtils() {}
+
+    /**
+     * Target-factor normalizes a {@link RealMatrix} vector in-place given target factors..
```
the word "vector" seems out of place

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * Target-factor normalizes a {@link RealMatrix} vector in-place given target factors..
+     */
+    static void factorNormalize(final RealMatrix input, final double[] targetFactors) {
```
Needs an access modifier

samuelklee replied:
i think package-protected is fine, no?

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+     * @param doFactorNormalization if true, perform factor normalization (set to false for normalizing normals in a PoN that have already been factor normalized)
+     * @return never {@code null}
+     */
+    static PCATangentNormalizationResult tangentNormalize(final PCACoveragePoN pon, final ReadCountCollection profile, final boolean doFactorNormalization, final JavaSparkContext ctx) {
```
access modifier

samuelklee replied:
same

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+     * (since target variances are considered part of {@link PCACoveragePoN}, this cannot be done simply using
+     * {@link CoveragePanelOfNormals#normalizeNormalsInPoN} since the PoN is not fully initialized).
+     */
+    static PCATangentNormalizationResult tangentNormalizeNormalsInPoN(final ReadCountCollection normalizedCounts,
```
access modifier

samuelklee replied:
same

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    /**
+     * Returns a target-factor--normalized {@link ReadCountCollection} given a {@link PCACoveragePoN}..
```
factor--normalized (two dashes) -> factor-normalized (one dash)

samuelklee replied:
done.  i'm not happy with the use of both "target factor normalized" and "factor normalized" along with inconsistent hyphenation, but i guess there are worse things

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * Calculate the beta-hats that best fit case read counts given the panel of normals.
+     *
+     * @param normalsInverse the log-normalized or reduced-panel pseudoinverse from a panel of normals
```
Call it `normalsPseudoinverse`

samuelklee replied:
done

samuelklee replied:
also not happy about PInverse vs Pseudoinverse (PInverse sounds like the inverse of matrix P) either, but i'll live

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+     *
+     * @param normalsInverse the log-normalized or reduced-panel pseudoinverse from a panel of normals
+     * @param input a {@code TxS} matrix where {@code T} is the number of targets and {@code S} the number of count groups (e.g. case samples).
+     * @return never {@code null} a matrix of {@code NxS} dimensions, where N is the number of samples in
```
A matrix of NxS dimensions -> an NxS matrix

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    public static RealMatrix calculateBetaHats(final RealMatrix normalsInverse, final RealMatrix input, final double epsilon) {
+        Utils.nonNull(normalsInverse, "Normals inverse matrix cannot be null.");
+        Utils.nonNull(input, "Input counts cannot be null.");
+        if (Double.isNaN(epsilon) || epsilon <= 0) {
```
Utils.validateArg

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * Applies tangent normalization.
+     * <p>
+     *     The input row order is suppose to match the panels target order.
```
supposed; panel's

samuelklee replied:
"The input row order should match the panel's target order."

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+     * @return never {@code null}.
+     */
+    private static RealMatrix tangentNormalize(final RealMatrix normals, final RealMatrix input, final RealMatrix betaHats) {
+        if (input.getColumnDimension() != betaHats.getColumnDimension()) {
```
Utils.validateArg x 2

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+        final RowMatrix projectionTDistMat = betahatDistMat.multiply(panelTLocalMat);
+        final RealMatrix projection = SparkConverter.convertSparkRowMatrixToRealMatrix(projectionTDistMat, tangentNormalizationInputCounts.transpose().getRowDimension()).transpose();
+
+        // Subtract the cases from the projection
```
cases from the projection -> projection from the cases

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    /**
+     * Prepares the data to perform tangent factor normalization.
```
old terminology?

samuelklee replied:
done

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5File.java:
```diff
@@ -277,6 +280,54 @@ public double readDouble(final String fullPath) {
+            return result;
+        }
+        final RealMatrix transpose = result.transpose();
+        if (!expectedRowCount.test(transpose.getRowDimension())) {
```
Why does this need to be an `IntPredicate` instead of an `int`?  Is this ever passes a predicate that is not checking for equality with an expected value?

samuelklee replied:
not sure, but this is over in public now...

samuelklee replied:
actually, i take that back, these methods have to be moved back to the PoN class.  apparently the check for the dimensions of the reduced panel matrices only requires that the number of samples is less than or equal to that returned by `getPanelSampleNames()`.  again, not sure why this is, but i'll leave it as is for now.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/pon/PoNTestUtils.java:
```diff
@@ -130,24 +113,29 @@ public static void assertEquivalentPoN(final File left, final File right) {
+        Assert.assertEquals(leftPoN.getRawTargets(), rightPoN.getRawTargets());
+        Assert.assertEquals(leftPoN.getPanelTargets(), rightPoN.getPanelTargets());
+
+        PoNTestUtils.assertEqualsDoubleArrays(leftPoN.getTargetFactors(), rightPoN.getTargetFactors());
```
Doesn't Assert.assertEquals(Object, Object) recognize if its arguments are `double[]`s and proceed accordingly?

samuelklee replied:
does not allow for a tolerance, so i guess that's why this method exists

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNUnitTest.java:
```diff
@@ -0,0 +1,232 @@
+     * @throws IOException if some IO exception occurred.
+     * @throws NumberFormatException if some of the input lines in {@code file} cannot be converted into a {@code double}.
+     */
+    private List<Double> readDoubleLines(final File file) throws IOException {
```
Move to `ParamUtils`?

samuelklee replied:
eh...seems like a method just for test files, is that the appropriate place?  i think all our production code should go through TableReader/TableWriter or something more robust, but other tests could make use of these methods.  punting for now

davidbenjamin commented:
Done reviewing "Major refactoring of PCA coverage PoN".  Nice work!  Looks a lot better.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5PoN.java:
```diff
@@ -35,60 +35,65 @@
+     *===============================================================================================================*/
 
-    private final static String VERSION_GROUP_NAME = "/version";
+    public static final String TARGET_FACTORS_GROUP_NAME = "/target_factors";
```
Was there are reason for making these public? Also I believe you moved this file in a different location in a later commit

samuelklee replied:
the original access modifiers weren't changed (it might be confusing because i shuffled the order of the fields around)

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CalculatePulldownPhasePosteriors.java:
```diff
@@ -86,7 +87,7 @@ public Object doWork() {
 
         //load allelic-bias panel of normals if provided
         final AllelicPanelOfNormals allelicPON =
```
allelicPON -> allelicPoN

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CalculatePulldownPhasePosteriors.java:
```diff
@@ -86,7 +87,7 @@ public Object doWork() {
         //load allelic-bias panel of normals if provided
         final AllelicPanelOfNormals allelicPON =
-                allelicPONFile != null ? new AllelicPanelOfNormals(allelicPONFile) : AllelicPanelOfNormals.EMPTY_PON;
+                allelicPONFile != null ? AllelicPanelOfNormals.read(allelicPONFile) : AllelicPanelOfNormals.EMPTY_PON;
```
allelicPON -> allelicPoN

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/exome/allelefraction/AlleleFractionModellerUnitTest.java:
```diff
@@ -167,18 +168,18 @@ private void testMCMC(final double meanBiasSimulated, final double biasVarianceS
 
-        final AllelicPanelOfNormals allelicPONNormal = new AllelicPanelOfNormals(ALLELIC_PON_NORMAL_FILE);
-        final AllelicPanelOfNormals allelicPONWithBadSNPs = new AllelicPanelOfNormals(ALLELIC_PON_WITH_BAD_SNPS_FILE);
+        final AllelicPanelOfNormals allelicPoNNormal = new AllelicPanelOfNormals(new AllelicCountCollection(ALLELIC_PON_NORMAL_COUNTS_FILE));
```
Should this be allelicPonNormal according to Google conventions doc?

samuelklee replied:
Yup, but will pass as mentioned above.

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/exome/allelefraction/AlleleFractionModellerUnitTest.java:
```diff
@@ -167,18 +168,18 @@ private void testMCMC(final double meanBiasSimulated, final double biasVarianceS
-                {sampleWithBadSNPs, allelicPONWithBadSNPs, minorFractionExpectedInMiddleSegmentNormal}
+                {sampleNormal, allelicPoNNormal, minorFractionExpectedInMiddleSegmentNormal},
+                {sampleWithBadSNPs, allelicPoNNormal, minorFractionExpectedInMiddleSegmentWithBadSNPsAndNormalPON},
+                {sampleWithEvent, allelicPoNNormal, minorFractionExpectedInMiddleSegmentWithEvent},
```
Similar here

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/utils/param/ParamUtilsUnitTest.java:
```diff
@@ -261,7 +261,7 @@ public void testLog2OfNegativeNumber() {
     @Test
     public void testWriteAndReadDoublesToFile() {
-        final double [] testData = {0.0, 1.234552345, Math.PI, Double.NaN, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, -3.91};
+        final double[] testData = new double[]{0.0, 1.234552345, Math.PI, Double.NaN, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, -3.91};
```
space after `double[]`

samuelklee replied:
Good catch!  I'll go ahead and change other instances of this as well.

asmirnov239 commented on src/test/resources/org/broadinstitute/hellbender/tools/exome/allelic-pon-test-pon-counts-normal.tsv:
```diff
@@ -0,0 +1,1001 @@
+1	352	1257	1192	50
+1	353	1262	1228	50
+1	354	1490	1057	50
+1	355	1141	1382	49
```
Why are the few of the values in the last column 49?

samuelklee replied:
Good catch.  This is because in the python script that simulated this pon there is an a read-count threshold---it doesn't include a sample if the read counts it has at the site are below the threshold.

This `NUM_SAMPLES` column isn't currently used, so I'll remove it from the test files.  However, it might actually be good to store this in future versions of the allelic pon (in case someone is interested in seeing how many samples actually contained each site).  I'll file an issue for this.

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,94 @@
+    protected List<File> inputFiles = new ArrayList<>();
+
+    @Argument(
+            doc = "Output HDF5 file.  The allelic panel of normals will be written to this file (which will be overwritten, if it exists).",
```
Can this be wrapped to 2 lines?

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,94 @@
+    protected File outputTSVFile = null;
+
+    @Argument(
+            doc = "Site frequency threshold.  Sites that appear in a fraction of the samples that is strictly less than this threshold are not included in the panel.",
```
Can this be wrapped to 2 lines?

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,94 @@
+            fullName = SITE_FREQUENCY_LONG_NAME,
+            optional = true
+    )
+    protected double siteFrequency = 0.25;
```
Would it be more clear to make it something like `minSiteFrequency`, or `siteFrequencyThreshold`

samuelklee replied:
done throughout

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/AllelicCNV.java:
```diff
@@ -116,7 +116,7 @@
             optional = true
     )
-    protected File allelicPONFile;
+    protected File allelicPoNFile;
```
Should this be `allelicPonFile`? I don't know how much we follow the Google conventions...

samuelklee replied:
You're right.  We have been a little lax on following the camel-case conventions exactly, but no time like the present to start being more strict!

samuelklee replied:
Hmm, on second thought, there are violations everywhere, so that changing this to follow the convention might actually make it stick out more...  I think I will fix `allelicPON` -> `allelicPoN` as you suggest, though.

samuelklee replied:
also changed `PON` -> `PoN` where appropriate

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,94 @@
+        validateArguments();
+
+        logger.info("Starting allelic panel of normals creation...");
+        final AllelicPanelOfNormals allelicPoN = new AllelicPanelOfNormalsCreator(inputFiles).create(siteFrequency);
```
I noticed in some other command line classes, the total run times are recorded. Is it worth doing it here?

samuelklee replied:
Good observation.  However, since it doesn't take very long to create the allelic pon (<10s for 100s of exomes, <5 minutes for ~100 WGS), I will pass on it for now.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,94 @@
+package org.broadinstitute.hellbender.tools.exome;
+
+import org.broadinstitute.hellbender.cmdline.*;
```
is there a reason to do a wildcard import?

samuelklee replied:
I think we usually go with whatever intelliJ's "Optimize imports" gives us.

droazen commented:
Recommend rebasing this branch onto @akiezun's now-merged change to pull HDF5 down from maven central sooner rather than later, so that if there are issues @akiezun will be there to help.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+     * @return  MLE global hyperparameter values alpha and beta
+     */
+    public HyperparameterValues getGlobalHyperparameterValues() {
+        throwExceptionIfPoNIsEmpty();
```
`throwExceptionIfPonIsEmpty()`

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+        logger.info("Allelic panel of normals initialized.");
+    }
+
+    private void throwExceptionIfPoNIsEmpty() {
```
`throwExceptionIfPonIsEmpty()`

samuelklee replied:
see above

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+    }
+
+    //returns a list of the sites in siteToHyperarameterPairMap that are sorted by SimpleInterval (contigs are sorted by lexicographical order)
+    private static List<Map.Entry<SimpleInterval, HyperparameterValues>> collectSortedMapEntries(final Map<SimpleInterval, HyperparameterValues> siteToHyperparameterPairMap) {
```
Should this be line-wrapped? What is the policy we use for line-wrapping in general?

ldgauthier replied:
Is there a code style/standards doc for gatk4? Picard has one but gatk3 didn't really use one.

samuelklee replied:
I think we are supposed to adhere to the Google Java Guide (https://google.github.io/styleguide/javaguide.html), but we are a little lenient about some of the guidelines.  There are also a few style idiosyncrasies unique to gatk-protected that have evolved and we've stuck to for consistency.

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+
+/**
+ * Represents the panel of normals used for allele-bias correction.  See docs/CNVs/CNV-methods.pdf.
+ *
```
Maybe some small description of what the the hyperparameters in the panel are?

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    private static boolean isHDF5File(final File inputFile) {
```
Could this be delegated to one of the classes that deals with HDF5 specifically? Seems like some other classes might want to use it

samuelklee replied:
I agree, but I'm a little hesitant to as it's somewhat hacky...  I'll revisit this after I rebase onto the recent HDF5 changes.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+        try {
+            final List<String> commentLines = FileUtils.readLines(inputFile).stream().filter(l -> l.contains(TableUtils.COMMENT_PREFIX)).collect(Collectors.toList());
+            final List<String> globalAlphaCommentLines = commentLines.stream().filter(l -> l.contains(GLOBAL_ALPHA_COMMENT_STRING)).collect(Collectors.toList());
+            final List<String> globalBetaCommentLines = commentLines.stream().filter(l -> l.contains(GLOBAL_BETA_COMMENT_STRING)).collect(Collectors.toList());
```
Could these lines be line-wrapped?

samuelklee replied:
done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    private static HyperparameterValues parseGlobalHyperparameterValues(final File inputFile) {
```
Some description of the method?

samuelklee replied:
we don't typically require javadoc for private methods, but I will add some comments

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    private void initializeSiteToHyperparameterPairMap(final AllelicCountCollection counts) {
```
Description of the method

samuelklee replied:
will add comments

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormalsCreator.java:
```diff
@@ -0,0 +1,82 @@
+                    numberOfSamplesMap.put(site, new MutableInt(1));
+                } else {
+                    numberOfSamplesAtSite.increment();
+                }
```
Just a question! Is this the common way in our codebase to do word count? It seems like it's one of the things CNV pipeline does often, and I read it's one of the faster ways - so I just want to know in the future. Thanks!

samuelklee replied:
I'm not actually sure!  In general, I wouldn't say we try to favor particular implementations of specific algorithms.  As long as your implementation is 1) working and correct, 2) understandable and maintainable, 3) not terribly inefficient, we'll probably allow it (although the passing threshold for each is up to the PR reviewer).  Of course, if certain bits of code are crucial to performance then we'll optimize where necessary, but that can happen in subsequent PRs.

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormalsIntegrationTest.java:
```diff
@@ -0,0 +1,119 @@
+        return new Object[][]{
+                {new String[]{"--" + SITE_FREQUENCY_LONG_NAME, "0"}},
+                {new String[]{"--" + SITE_FREQUENCY_LONG_NAME, "10"}},
+                {new String[]{"--" + SITE_FREQUENCY_LONG_NAME, "-1"}}
```
Space after `[]`

samuelklee replied:
This is another one that we are inconsistent on (actually it looks like we seem prefer it without the space).  Perhaps a good guideline would be to keep things consistent within each class, at least.

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormalsIntegrationTest.java:
```diff
@@ -0,0 +1,119 @@
+    public Object[][] dataSiteFrequency() {
+        return new Object[][]{
+                {new String[]{"--" + SITE_FREQUENCY_LONG_NAME, "0.5"}, PON_EXPECTED_SITE_FREQUENCY_50_HDF5, PON_EXPECTED_SITE_FREQUENCY_50_TSV},
+                {new String[]{"--" + SITE_FREQUENCY_LONG_NAME, "0.75"}, PON_EXPECTED_SITE_FREQUENCY_75_HDF5, PON_EXPECTED_SITE_FREQUENCY_75_TSV}
```
Space after `[]`

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormalsUnitTest.java:
```diff
@@ -32,34 +33,34 @@
+        final AllelicPanelOfNormals allelicPoN = new AllelicPanelOfNormals(new AllelicCountCollection(ALLELIC_PON_NORMAL_COUNTS_FILE));
 
         final SimpleInterval firstSite = new SimpleInterval("1", 1, 1);  //all sites in PON are from chr1
         final SimpleInterval siteNotInPON = new SimpleInterval("2", 1, 1);  //all sites in PON are from chr1
```
`siteNotInPON `-> `siteNotInPoN `

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormalsUnitTest.java:
```diff
@@ -32,34 +33,34 @@
-        final double meanBias = allelicPON.getMLEMeanBias();
-        final double biasVariance = allelicPON.getMLEBiasVariance();
+        final double alphaNotInPON = allelicPoN.getAlpha(siteNotInPON);
+        final double betaNotInPON = allelicPoN.getBeta(siteNotInPON);
```
PON -> PoN

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormalsCreatorUnitTest.java:
```diff
@@ -0,0 +1,51 @@
+
+    @DataProvider(name = "dataCreate")
+    public Object[][] dataCreate() throws IOException {
+        return new Object[][]{
```
Space after `[]`

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/HDF5AllelicPoNUtils.java:
```diff
@@ -0,0 +1,155 @@
+
+    private static final Map<String, Integer> snpColumnToPoNIndex = ImmutableMap.of(
+            AllelicCountTableColumn.CONTIG.toString(), 0, AllelicCountTableColumn.POSITION.toString(), 1
+    );
```
Move `);` to previous line

samuelklee replied:
done here (and also in corresponding place in `HDF5PCACoveragePoN`

asmirnov239 commented:
@samuelklee Finished reviewing the 4 commits assigned to me. The code looks very well written. Awesome job! Home my comments help

samuelklee commented:
@asmirnov239 @davidbenjamin back to you guys.  i split up some of the post-review changes into separate commits to hopefully make it easier for you guys to parse.  also went ahead and rebased (which was a little messy); changes there are also in their own commit.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/exome/CreateAllelicPanelOfNormals.java:
```diff
@@ -70,6 +75,11 @@
+            throw new UserException.HardwareFeatureException("Cannot load the required HDF5 library. " +
+                    "HDF5 is currently supported on x86-64 architecture and Linux or OSX systems.");
+        }
+
```
This code snippet is repeated in `AllelicCNV` and `CalculatePulldownPhasePosteriors` classed. Could we move it to some Utils class?

samuelklee replied:
I agree that it is awkward to repeat.  See discussion in #622.  Maybe @lbergelson can advise?

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/pon/allelic/AllelicPanelOfNormals.java:
```diff
@@ -317,6 +320,47 @@ private static HyperparameterValues parseGlobalHyperparameterValues(final File i
+            nextIterationLogLikelihood = AlleleFractionLikelihoods.logLikelihoodForAllelicPanelOfNormals(meanBias, biasVariance, counts);
+            logger.info(String.format("Iteration %d, model log likelihood = %.3f.", iteration, nextIterationLogLikelihood));
+            iteration++;
+        } while (iteration < AlleleFractionInitializer.MAX_ITERATIONS &&
```
Should the method give some kind of warning if it didn't converge during first `AlleleFractionInitializer.MAX_ITERATIONS` iterations?

samuelklee replied:
This is a good point, I'll file an issue to add these sort of warnings.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5File.java:
```diff
@@ -1,1052 +0,0 @@
-package org.broadinstitute.hellbender.utils.hdf5;
```
So do I understand it correctly, that all this code has been moved to the new hdf5 repo?

samuelklee replied:
Yes, it should all be in https://github.com/broadinstitute/hdf5-java-bindings

In my opinon, there are a few methods that got stranded in this repo (like some things you pointed out, like checking that the library is loaded or whether a file can be read as HDF5), as well as a few PoN-specific things that made their way into that repo that could be removed.  But we can sort this out later.

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNUnitTest.java:
```diff
@@ -37,7 +43,15 @@
+        if (!new HDF5Library().load(null)) {  //Note: passing null means using the default temp dir.
+            throw new UserException.HardwareFeatureException("Cannot load the required HDF5 library. " +
+                    "HDF5 is currently supported on x86-64 architecture and Linux or OSX systems.");
+        }
```
Same as above

asmirnov239 commented:
@samuelklee  Finished reviewing the latest commits addressing my comments + "Resolving rebase merge conflicts..." commit. I left a couple of comments but other than that everything looks good to me!

samuelklee commented:
actually, @mbabadi, can you look over the commits addressing @davidbenjamin's comments and the rebase commit?  just a quick pass should be fine.

samuelklee commented:
one more commit to take care of incorrect tests in CreatePanelOfNormalsIntegrationTest.

mbabadi commented:
@samuelklee sure, will do.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -186,7 +187,8 @@ public double getVersion() {
         if (values.length != targetNames.get().size()) {
-            throw new GATKException(String.format("Wrong number of elements in the target factors recovered from file '%s': %d != %d", file.getFile(), values.length, targetNames.get().size()));
+            throw new GATKException(String.format("Wrong number of elements in the target factors recovered " +
+                    "from file '%s': %d != %d", file.getFile(), values.length, targetNames.get().size()));
```
It would be more informative to replace "%d != %d" in the exception message with "(from file) %d != (number of targets) %d"

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -195,7 +197,8 @@ public double getVersion() {
         final double[] values = file.readDoubleArray(TARGET_VARIANCES_PATH);
         if (values.length != panelTargets.get().size()) {
-            throw new GATKException(String.format("wrong number of elements in the target variances recovered from file '%s': %d != %d", file.getFile(), values.length, panelTargets.get().size()));
+            throw new GATKException(String.format("wrong number of elements in the target variances recovered " +
```
same as above. Also, "wrong" -> "Wrong"

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -365,14 +368,16 @@ public static void write(final File outFile,
     private static List<Target> renderPoNTargets(final String[][] values, final List<String> targetNamesToRender, final HDF5File reader) {
         if (values.length != targetNamesToRender.size()) {
-            throw new GATKException(String.format("Wrong number of elements in the targets recovered from file '%s': %d != %d", reader.getFile(), values.length, targetNamesToRender.size()));
+            throw new GATKException(String.format("Wrong number of elements in the targets recovered " +
```
same as above.

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoN.java:
```diff
@@ -365,14 +368,16 @@ public static void write(final File outFile,
         for (int i= 0; i < values.length; i++) {
             if (values[i].length != numTargetCols) {
-                throw new GATKException(String.format("Wrong number of column elements in the targets recovered from file '%s': %d != %d", reader.getFile(), values[i].length, numTargetCols));
+                throw new GATKException(String.format("Wrong number of column elements in the targets recovered " +
```
same as above.

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/HDF5PCACoveragePoNCreationUtils.java:
```diff
@@ -455,7 +454,8 @@ static ReadCountCollection removeColumnsWithExtremeMedianCounts(final ReadCountC
         // Determine kept and dropped column sets.
         final Set<String> columnsToKeep = new HashSet<>(readCounts.columnNames().size());
-        final Map<String, Double> columnsToDrop = new LinkedHashMap<>((int) (extremeColumnMedianCountPercentileThreshold * 4.0 / 100.0));
+        final int initialMapSize = ((int) (2. * extremeColumnMedianCountPercentileThreshold / 100.)) * readCounts.columnNames().size();
```
Any reason why 4 -> 2?

samuelklee replied:
in response to a comment by @davidbenjamin.  the original value of 4 was probably incorrect or overkill.  we drop extreme values below the percentile threshold and above 1 - the threshold, so the number of values is roughly twice the threshold.  i'm not sure if this actually makes a difference in performance, though (or if being off by one and making the initial map slightly too small will hurt performance).

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -169,50 +168,46 @@ private static PCATangentNormalizationResult tangentNormalizeNonSpark(final Read
-        if (normals.getColumnDimension() != betaHats.getRowDimension()) {
-            throw new IllegalArgumentException(String.format("beta-hats component count (%d) does not match the number of samples in the PoN (%d)", normals.getRowDimension(), normals.getColumnDimension()));
-        }
+        Utils.validateArg(input.getColumnDimension() == betaHats.getColumnDimension(), String.format("the input count column count (%d) does not match the number of columns in the beta-hats (%d)", input.getColumnDimension(), betaHats.getColumnDimension()));
```
line break please

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -169,50 +168,46 @@ private static PCATangentNormalizationResult tangentNormalizeNonSpark(final Read
-            throw new IllegalArgumentException(String.format("beta-hats component count (%d) does not match the number of samples in the PoN (%d)", normals.getRowDimension(), normals.getColumnDimension()));
-        }
+        Utils.validateArg(input.getColumnDimension() == betaHats.getColumnDimension(), String.format("the input count column count (%d) does not match the number of columns in the beta-hats (%d)", input.getColumnDimension(), betaHats.getColumnDimension()));
+        Utils.validateArg(normals.getColumnDimension() == betaHats.getRowDimension(), String.format("beta-hats component count (%d) does not match the number of samples in the PoN (%d)", normals.getRowDimension(), normals.getColumnDimension()));
```
line break please

samuelklee replied:
done

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/pon/coverage/pca/PCATangentNormalizationUtils.java:
```diff
@@ -250,7 +245,7 @@ private static PCATangentNormalizationResult tangentNormalizeSpark(final ReadCou
         final RowMatrix projectionTDistMat = betahatDistMat.multiply(panelTLocalMat);
         final RealMatrix projection = SparkConverter.convertSparkRowMatrixToRealMatrix(projectionTDistMat, tangentNormalizationInputCounts.transpose().getRowDimension()).transpose();
```
line break please

mbabadi replied:
line break please

samuelklee replied:
done

mbabadi commented:
I am done with the review of the last four commits, good job -- nothing much to comment on (based on looking at the last four) except for a few aesthetic things.

As a suggestion, I think it is a good idea to move `removeColumnsWithTooManyZeros`, `removeTargetsWithTooManyZeros`, `removeColumnsWithExtremeMedianCounts`, `imputeZeroCountsAsTargetMedians`, and `truncateExtremeCounts` from `HDF5PCACoveragePoNCreationUtils` to `ReadCountCollectionUtils`. The use case of these read count collection filters now exceeds the domain of HD5 PCA PoN creation. I'm using (some of) these methods in `TargetCoverageSexGenotyper` and the new probabilistic target coverage model. If you prefer to keep `ReadCountCollectionUtils` just for I/O methods (in which case it could be called `ReadCountCollectionIOUtils`), then we can create a separate class for the filters with a more generic name. Here's my extended `ReadCountCollectionUtils`:
https://github.com/broadinstitute/gatk-protected/blob/804ff6582f092efc3367a26697fb13ef0215d8c9/src/main/java/org/broadinstitute/hellbender/tools/exome/ReadCountCollectionUtils.java
https://github.com/broadinstitute/gatk-protected/blob/804ff6582f092efc3367a26697fb13ef0215d8c9/src/test/java/org/broadinstitute/hellbender/tools/exome/ReadCountCollectionUtilsUnitTest.java

samuelklee commented:
thanks!  will file an issue to move the filtering methods out, but what you did looks good to me,
